File: android/src/app/src/main/java/com/example/agitodo/Dropbox.java
package com.example.agitodo;

import java.util.ArrayList;
import java.util.List;

import org.apache.http.message.BasicNameValuePair;
import org.apache.http.NameValuePair;

import android.net.Uri;

public class Dropbox {
    static final String METADATA_ENDPOINT = "https://api.dropbox.com/1/metadata/sandbox";
    static final String FILE_GET_ENDPOINT = "https://api-content.dropbox.com/1/files/sandbox";
    static final String FILE_PUT_ENDPOINT = "https://api-content.dropbox.com/1/files_put/sandbox";

    static Logger log = new Logger();

    String access_token;

    public Dropbox(String access_token) {
        this.access_token = access_token;
    }

    public Request.Response file_get(String path) {
        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("Authorization", "Bearer "
                + access_token));

        return new Request(Request.GET, Uri.parse(FILE_GET_ENDPOINT + path),
                headers).send(null);
    }

    public Request.Response file_put(String path, String body) {
        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("Authorization", "Bearer "
                + access_token));
        headers.add(new BasicNameValuePair("Content-Type",
                "text/plain; charset=utf-8"));

        return new Request(Request.POST, Uri.parse(FILE_PUT_ENDPOINT + path),
                headers).send(body);
    }
}

File: www/dropbox.js
var url = require("url");
var path = require("path");

var request = require("./request");
var oauth = require("./oauth");

var METADATA_ENDPOINT = "https://api.dropbox.com/1/metadata/sandbox";
var FILE_GET_ENDPOINT = "https://api-content.dropbox.com/1/files/sandbox";
var FILE_PUT_ENDPOINT = "https://api-content.dropbox.com/1/files_put/sandbox";

function token(session, callback, code, refresh) {
  if (typeof code !== "string" || !code) {
    return callback ? callback() : undefined;
  }

  oauth.token(session, oauth.SERVICE.DROPBOX, code, refresh, callback);
}

function metadata(session, callback, access_token, filePath) {
  if (typeof access_token !== "string" || typeof filePath !== "string") {
    return callback ? callback() : undefined;
  }

  while (filePath.charAt(0) === "/") {
    filePath = filePath.slice(1);
  }

  var headers = {
    "Authorization": "Bearer " + access_token
  };

  var u = url.parse(METADATA_ENDPOINT);

  request("GET", u.hostname, path.join(u.pathname, filePath), headers, null,
    null, callback);
}

function file_get(session, callback, access_token, filePath) {
  if (typeof access_token !== "string" || typeof filePath !== "string") {
    return callback ? callback() : undefined;
  }

  while (filePath.charAt(0) === "/") {
    filePath = filePath.slice(1);
  }

  var headers = {
    "Authorization": "Bearer " + access_token
  };

  var u = url.parse(FILE_GET_ENDPOINT);

  request("GET", u.hostname, path.join(u.pathname, filePath), headers, null,
    null, callback);
}

function file_put(session, callback, access_token, filePath, body) {
  if (typeof access_token !== "string" || typeof filePath !== "string" ||
    typeof body !== "string") {
    return callback ? callback() : undefined;
  }

  while (filePath.charAt(0) === "/") {
    filePath = filePath.slice(1);
  }

  var headers = {
    "Authorization": "Bearer " + access_token,
    "Content-Type": "text/plain; charset=utf-8"
  };

  var u = url.parse(FILE_PUT_ENDPOINT);

  request("POST", u.hostname, path.join(u.pathname, filePath), headers, null,
    new Buffer(body,
      "utf8"), callback);
}

module.exports = {
  token: token,
  metadata: metadata,
  file_get: file_get,
  file_put: file_put
};
